### PR TITLE
Add debug logging option

### PR DESCRIPTION
**Pull request by dblandin**
This PR adds a configuration option for debug logging in order to shake out some difficult to diagnose analysis issues.

@codeclimate/review :mag_right:

### CODE REVIEW

File: lib/cc/engine/duplication.rb [hunk 1]
```diff
@@ -30,6 +30,10 @@ def initialize(directory:, engine_config:, io:)
       def run
         Dir.chdir(directory) do
           languages_to_analyze.each do |language|
+            if engine_config.debug?
+              $stderr.puts("Analyzing language #{language}")
```
Comment (pbrisbin) [inline lib/cc/engine/duplication.rb]: This seems to be the only log line that can't use `Reporter#debug` and has to duplicate the logic. Thoughts on just ditching it? Seems low value to me.

**Comment by pbrisbin** [comment]
LGTM. Note that to do CLI testing, you'll need to make a change there to display engine STDERR even in the successful case (it doesn't do this currently).

File: lib/cc/engine/duplication.rb (hunk 1 above)
Comment (dblandin) [inline lib/cc/engine/duplication.rb]: Ditching it sounds good to me.

**Comment by dblandin** [comment]
> Note that to do CLI testing, you'll need to make a change there to display engine STDERR even in the successful case (it doesn't do this currently).

Right. I've been running the engine directly with `docker run` to test. For our purposes debugging some specific issues, I think it's fine if it doesn't show up during CLI runs. We just need to see the engine stderr on codeclimate.com.

Do you think this is useful enough generally to make that CLI change?

File: lib/cc/engine/analyzers/reporter.rb [hunk 2]
```diff
@@ -26,12 +30,20 @@ def process_files
...
 
+          processed_files_count = 0
+
           pool.run do |file|
+            debug("Processing file: #{file}")
+
             sexp = language_strategy.run(file)
             process_sexp(sexp)
+
+            processed_files_count += 1
```
Comment (GordonDiggs) [inline lib/cc/engine/analyzers/reporter.rb]: Do you need anything here to handle the concurrency and incrementing this number?

Comment (dblandin) [inline lib/cc/engine/analyzers/reporter.rb]: I don't think so. The join below should block until all threads return.

Comment (GordonDiggs) [inline lib/cc/engine/analyzers/reporter.rb]: Right but this block is executed inside the thread right? So more than one thread could try to increment this at once

**Comment by pbrisbin** [comment]
> Do you think this is useful enough generally to make that CLI change?

Nope, just letting you know about a gotcha that I've hit before. Whoever thinks it's useful should make that change.

File: lib/cc/engine/analyzers/reporter.rb (hunk 2 above)
Comment (dblandin) [inline lib/cc/engine/analyzers/reporter.rb]: Gotcha, I suppose that's true! I'll wrap the increment with a mutex.

Comment (GordonDiggs) [inline lib/cc/engine/analyzers/reporter.rb]: You could also use this class from concurrent-ruby I think http://ruby-concurrency.github.io/concurrent-ruby/Concurrent/AtomicFixnum.html

Comment (dblandin) [inline lib/cc/engine/analyzers/reporter.rb]: Nice! That's handy!

**Pull request merged**